README.md
:warning: **This project is no longer maintained.** :warning:<br>
**Due to my SkyBell HD failing and the difficulty of obtaining a replacement I have replaced it with a Logitech Circle View Doorbell.**

<p align="center">
  <a href="https://github.com/homebridge/homebridge/wiki/Verified-Plugins"><img src="https://raw.githubusercontent.com/wiki/thoukydides/homebridge-skybell/homebridge-skybell.png" height="200"></a>
</p>
<span align=center>

# homebridge-skybell

[![npm](https://badgen.net/npm/v/homebridge-skybell)](https://www.npmjs.com/package/homebridge-skybell)
[![npm](https://badgen.net/npm/dt/homebridge-skybell)](https://www.npmjs.com/package/homebridge-skybell)
[![npm](https://badgen.net/npm/dw/homebridge-skybell)](https://www.npmjs.com/package/homebridge-skybell)
[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)
[![certified-hoobs-plugin](https://badgen.net/badge/HOOBS/certified/yellow)](https://plugins.hoobs.org)

SkyBell HD Wi-Fi video doorbell plugin for [Homebridge](https://github.com/nfarina/homebridge).

</span>

SkyBell is a trademark owned by [SkyBell Technologies, Inc](http://www.skybell.com/).

## Installation

1. Install [FFmpeg](https://www.ffmpeg.org/) (with `libx264` and `libopus` codecs enabled) using: `sudo apt-get install ffmpeg`
1. Install this plugin using: `npm install -g homebridge-skybell`
1. Edit `config.json` and add the SkyBell platform (see example below).
1. Run [Homebridge](https://github.com/nfarina/homebridge).
1. Add extra camera accessories using the Home app:
   1. Click on the **+** (in the top-right corner of the screen) and select *Add Accessory*.
   1. Select *Don't Have a Code or Can't Scan?*
   1. Select the SkyBell HD camera accessory to be added.
   1. Ignore the warning about it being an *Uncertified Accessory* (which applies to all non-commercial accessories) and select *Add Anyway*.
   1. Enter the 8-digit setup code that Homebridge displayed when launched (possibly configured via a `pin` value in `config.json`).
   1. Select a room for the camera and motion sensor. They should be located in the same room, but not the *Default Room*.
 
### Example `config.json`
```JSON
{
    "platforms":
    [{
        "platform":     "SkyBell",
        "username":     "skybell@gmail.com",
        "password":     "Passw0rd!",
        "port":         47569,
        "secret":       "My webhooks secret"
    }]
}
```
The `username` (email address) and `password` should be set to the credentials used to log into the SkyBell HD app. The plugin will automatically create HomeKit accessories for all SkyBell HD (and SkyBell Trim Plus) doorbells associated with the account.

The `port` and `secret` values are optional. They should be omitted unless [webhooks](https://github.com/thoukydides/homebridge-skybell/wiki/Webhooks) are being used.

## Notes

### SkyBell API

SkyBell Technologies only share API details with selected third parties through their [SkyBell Connect API](http://www.skybell.com/skybell-connect/) program. Instead this plugin uses the [unpublished API](https://github.com/thoukydides/homebridge-skybell/wiki/Protocol-HTTPS) that the SkyBell HD app uses to communicate with the SkyBell cloud.

Without access to mobile push notifcations (Apple Push Notification service (APNs) or Google Cloud Messaging (GCM) for Android) this plugin has to poll the API for changes. It leaves 5 seconds between successive requests. Unfortunately, the SkyBell cloud servers appear to be very heavily loaded, with responses to each API request typically taking a few seconds.

### Button and Motion Events

By default button presses and motion events are detected by polling the SkyBell cloud for new video recordings being available. The doorbell only uploads the video after it has stopped recording so this typically results in a delay of several minutes between the event occurring and HomeKit being notified (and hence any automation being triggered).

The delay can be significantly reduced by snooping on the [packets between the doorbell and its cloud services](https://github.com/thoukydides/homebridge-skybell/wiki/Protocol-CoAP). If `tcpdump` can be run on the gateway/router between the doorbell and the internet then configure [webhooks](https://github.com/thoukydides/homebridge-skybell/wiki/Webhooks) and install the [doorbell packet sniffer](https://github.com/thoukydides/homebridge-skybell/wiki/Webhooks-Sniffer).

### Apple Home App

Apple's Home app (as of iOS 11) does not display *Doorbell* services; their characteristics cannot be viewed or controlled. For full functionality use one of the following apps:
* Matthias Hochgatterer's [Home](http://hochgatterer.me/home/) *(recommended)*.
* Elgato's [Eve](https://www.elgato.com/en/eve/eve-app) *(free)*.

However, Apple's Home app does generate rich notifications for both doorbell button presses and motion events. These appear with a small snapshot image from the camera. A larger image and live video stream can be viewed by sliding the notification left and selecting *View*. The doorbell button notification is enabled by default, but the motion event notification has to be explicitly enabled within the app.

### Snapshot Image

This plugin normally uses the *avatar* image that the SkyBell HD pushes to the SkyBell cloud once per hour. However, if a video was recorded more recently then its thumbnail will be used instead. The update time shown within the HomeKit app indicates when the image was requested, not when it was actually captured.

### Video and Audio Streams

Viewing the doorbell's video via HomeKit will normally establish a live stream showing the current camera image. However, within 30 minutes of a video being recorded (in response to a button press or motion event) this plugin will instead play that recording. In both cases the Home and Eve apps will badge the video as *LIVE*, so to make it clear when a recording is being played this plugin superimposes a caption at the top of recorded videos. Only recordings made after Homebridge is started are played.

Viewing the live stream from a SkyBell HD via HomeKit results in an *On demand* video being recorded, just as it does with the official SkyBell HD app. For some reason the SkyBell cloud terminates the live video stream after approximately 45 seconds when viewed via this plugin (instead of 5 minutes with the official SkyBell HD app). It is not obvious why this occurs.

This plugin currently only supports audio in a single direction: from the doorbell's microphone to HomeKit. This is because the audio streams in each direction use the same ports, but unfortunately FFmpeg (which is used by this plugin to transcode the audio and video) does not support sharing a single port between two SRTP streams.

## Changelog

All notable changes to this project will be documented in [CHANGELOG.md](CHANGELOG.md).

**The current version of this plugin requires Homebridge 1.0.0 or later. Use version 1.8.1 with older releases of Homebridge.**

## License

> ISC License (ISC)<br>Copyright © 2017-2021 Alexander Thoukydides
>
> Permission to use, copy, modify, and/or distribute this software for any purpose with or without fee is hereby granted, provided that the above copyright notice and this permission notice appear in all copies.
>
> THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

## [v2.1.5] - 2021-05-29
### Changed
* This project is no longer maintained.

## [v2.1.4] - 2021-03-02
### Fixed
* Re-use the existing bearer token when starting a call (live video stream) instead of repeating username/password authentication. ([#19])
* Ignore API responses that do not include all expected fields. ([#13] / [#22])

## [v2.1.3] - 2021-02-27
### Fixed
* Re-enabled playback of recordings with updated FFmpeg H.264 encoder options. ([#22])

## [v2.1.2] - 2021-02-27
### Fixed
* Work with change to private API in Homebridge 1.3.x (used to dynamically update supported video resolutions). ([#22] / [#23])
* Support more recent versions of FFmpeg that require `https` to be included in the whitelist protocols. ([#22])
### Removed
* Playback of recordings has been disabled (hopefully temporarily) due to an incompatibility with HomeKit. ([#22])

## [v2.1.1] - 2020-11-13
### Fixed
* Corrected callback handling when starting to stream the doorbell video. ([#21])

## [v2.1.0] - 2020-07-18
### Added
* Added a `User-Agent` header to all SkyBell API requests. ([#19])

## [v2.0.2] - 2020-05-24
### Added
* Added **[HOOBS Certified](https://plugins.hoobs.org/plugin/homebridge-skybell)** badge to the `README.md`. ([#15], [#17])

## [v2.0.1] - 2020-05-07
### Added
* Added **[Verified By Homebridge](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)** badge to the `README.md`.
* The homebridge API version is now checked at start-up.

## [v2.0.0] - 2020-05-03
### Fixed
* Now supports (and requires) homebridge version 1.0.0 or later. ([#16])

## [v1.8.1] - 2020-03-02
### Fixed
* Corrected the platform name in the configuration schema.

## [v1.8.0] - 2020-01-09
### Added
* Added a schema (`config.schema.json`) to allow editing of this plugin's configuration using [homebridge-config-ui-x](https://github.com/oznu/homebridge-config-ui-x). ([config.json])
* Created this `CHANGELOG.md` file.

## [v1.7.0] - 2019-12-29
### Fixed
* Corrected conversion of the LED colour between HomeKit and SkyBell representations. This previously failed when converting pure white to HomeKit, and for hues that were multiples of 60° when converting from HomeKit.

## [v1.6.0] - 2019-11-06
### Changed
* If an error occurs whilst retrieving an avatar image then the response is now inspected to check whether it contains an error message. This makes the log output more helpful when Amazon Simple Storage Service (S3) returns an error. ([#14])
### Fixed
* Fixed use of the `Buffer()` constructor to prevent the Node.js `DEP0005` deprecation warning.

## [v1.5.1] - 2019-01-28
### Removed
* Deleted a spurious file that was not intended to be committed.

## [v1.5.0] - 2018-07-08
### Fixed
* Replaced use of `pajk-lwip` by FFmpeg for scaling snapshot (avatar) images to the size requested by HomeKit. This allows use of Node.js version 10. ([#6])

## [v1.4.1] - 2017-12-31
### Fixed
* Corrected a typo that broke the handling of button press and motion event triggers in [v1.4.0].

## [v1.4.0] - 2017-12-29
### Changed
* FFmpeg is now only configured to transcode a video stream if either its resolution needs to be adjusted (to comply with HomeKit requirements) or an overlay is being added (to indicate a pre-recorded video). This should reduce latency and load on the Homebridge server.

## [v1.3.0] - 2017-12-28
### Fixed
* Switched from `lwip` to the `pajk-lwip` fork to fix compatibility with more recent versions of Node.js. ([#1])

## [v1.2.0] - 2017-12-27
### Added
* Use the thumbnail for the most recent video recording if it is newer than the current avatar image.

## [v1.1.0] - 2017-12-27
### Added
* When video streaming is started, if a new video was recorded within the last 30 minutes then that is played instead of live streaming from the SkyBell. When this happens the age of the video is added as an overlay.

## [v1.0.1] - 2017-12-26
### Changed
* Simplified parsing of error messages returned by the SkyBell API.

## [v1.0.0] - 2017-12-22
### Added
* Added webhooks that can be used to receive button press and motion detection event triggers from an external source. This allows latency to be reduced, e.g. by using IFTTT or network traffic sniffing to detect activity instead of polling the SkyBell API to check for a new video recording. ([Webhooks] / [Webhooks Sniffer])

## [v0.3.0] - 2017-12-19
### Removed
* Removed the unnecessary second `Motion Sensor` service that duplicated the functionality of the `Programmable Switch Event` characteristic for button presses.

## [v0.2.0] - 2017-12-19
### Removed
* Removed unnecessary handling of the `Microphone Mute` characteristic (since it is read-only).

## [v0.1.0] - 2017-12-19
* Initial version.

---

Copyright © 2017-2021 Alexander Thoukydides


accessory.js
// Homebridge plugin for SkyBell HD video doorbells
// Copyright © 2017-2021 Alexander Thoukydides

'use strict';

let SkyBellCameraStreamingDelegate = require('./camera');

let PlatformAccessory, CameraController;
let Accessory, Service, Characteristic, UUIDGen;
let VIDEO_DOORBELL, SINGLE_PRESS;

// Hard-coded characteristics
const MANUFACTURER = 'SkyBell Technologies, Inc.';
const TBD = '...';

// SkyBell only supports two concurrent calls
const MAX_STREAMS = 2;

// Length of time to signal motion (milliseconds)
const DURATION_MOTION = 10 * 1000;

// Length of time to ignore other sources when a trigger occurs (milliseconds)
const DURATION_SUPPRESS = 10 * 60 * 1000; // (10 minutes)

// Length of time to play recording instead of streaming live (milliseconds)
const DURATION_RECORDED = 30 * 60 * 1000; // (30 minutes)

// A Homebridge accessory for a SkyBell doorbell
module.exports = class SkyBellAccessory {
    
    // Initialise an accessory
    constructor(log, homebridge, skybellDevice, webhooks) {
        log("new SkyBellAccessory '" + skybellDevice.name + "': type='"
            + skybellDevice.cache.device.type + "'");
        this.log = log;
        this.skybellDevice = skybellDevice;
        this.name = skybellDevice.name;
    
        // Shortcuts to useful objects
        PlatformAccessory = homebridge.platformAccessory;
        CameraController = homebridge.hap.CameraController;
        Accessory = homebridge.hap.Accessory;
        VIDEO_DOORBELL = Accessory.Categories.VIDEO_DOORBELL;
        Service = homebridge.hap.Service;
        Characteristic = homebridge.hap.Characteristic;
        SINGLE_PRESS = Characteristic.ProgrammableSwitchEvent.SINGLE_PRESS;
        UUIDGen = homebridge.hap.uuid;

        // Create the accessory
        let uuid = UUIDGen.generate(skybellDevice.deviceId);
        this.accessory = new PlatformAccessory(this.name, uuid, VIDEO_DOORBELL);

        // Handle the identify request
        this.accessory.on('identify', this.identify.bind(this));

        // Add a hardware version characteristic
        // (not in x.y.z format required for Characteristic.HardwareRevision)
        this.informationService =
            this.accessory.getService(Service.AccessoryInformation);
        this.informationService
            .addCharacteristic(Characteristic.Version);
        this.informationService
            .setCharacteristic(Characteristic.Manufacturer, MANUFACTURER)
            .setCharacteristic(Characteristic.Model,
                               skybellDevice.cache.device.type)
            .setCharacteristic(Characteristic.SerialNumber, TBD)
            .setCharacteristic(Characteristic.FirmwareRevision, TBD)
            .setCharacteristic(Characteristic.Version, TBD);
                
        // Add a doorbell service
        this.doorbellService =
            this.accessory.addService(Service.Doorbell,
                                      this.name + ' doorbell');

        // Restrict the button to generating single press events only
        this.doorbellService
            .getCharacteristic(Characteristic.ProgrammableSwitchEvent)
            .setProps({ minValue: SINGLE_PRESS, maxValue: SINGLE_PRESS,
                        validValues: [SINGLE_PRESS] });
        
        // Add characteristics for the indoor and outdoor chimes
        this.doorbellService
            .addCharacteristic(Characteristic.Active)
            .on('set', this.setIndoorChime.bind(this))
            .displayName = 'Indoor Chime';
        this.doorbellService
            .addCharacteristic(Characteristic.Volume)
            .setProps({minValue: 0, maxValue: 99, minStep: 33,
                       validValues: [0, 33, 66, 99]})
            .on('set', this.setOutdoorChime.bind(this))
            .displayName = 'Outdoor Chime';
        
        // Add characteristics for controlling the LED
        this.doorbellService
            .addCharacteristic(Characteristic.Brightness)
            .setProps({minValue: 0, maxValue: 99, minStep: 33,
                       validValues: [0, 33, 66, 99]})
            .on('set', this.setLedIntensity.bind(this));
        this.doorbellService
            .addCharacteristic(Characteristic.Hue)
            .on('set', this.setLedHue.bind(this));
        this.doorbellService
            .addCharacteristic(Characteristic.Saturation)
            .on('set', this.setLedSaturation.bind(this));
        
        // Add a motion sensor service for motion detection
        this.motionSensorService =
            this.accessory.addService(Service.MotionSensor,
                                      this.name + ' motion sensor', 'motion');
        this.motionSensorService
            .addCharacteristic(Characteristic.Active)
            .on('set', this.setMotionPolicy.bind(this))
            .updateValue(false);

        // Add a speaker service
        this.speakerService =
            this.accessory.addService(Service.Speaker,
                                      this.name + ' camera speaker');
        this.speakerService
            .getCharacteristic(Characteristic.Mute)
            .setProps({perms: [Characteristic.Perms.READ,
                               Characteristic.Perms.NOTIFY]})
            .updateValue(true);
        this.speakerService.setHiddenService(true);
        
        // Add a microphone service
        this.microphoneService
            = this.accessory.addService(Service.Microphone,
                                        this.name + ' camera microphone');
        this.microphoneService
            .getCharacteristic(Characteristic.Mute)
            .setProps({perms: [Characteristic.Perms.READ,
                               Characteristic.Perms.NOTIFY]})
            .updateValue(false);
        this.microphoneService.setHiddenService(true);

        // Add a camera control service
        this.controlService = new Service.CameraControl(this.name + ' camera');
        this.controlService
            .getCharacteristic(Characteristic.On)
            .setProps({perms: [Characteristic.Perms.READ,
                               Characteristic.Perms.NOTIFY]})
            .updateValue(true);
        this.services = [this.controlService];

        // Add the camera controller services
        this.streamingDelegate =
            new SkyBellCameraStreamingDelegate(log, homebridge, skybellDevice);
        let options = {
            cameraStreamCount:  MAX_STREAMS,
            delegate:           this.streamingDelegate,
            streamingOptions:   this.streamingDelegate.getCodecParameters()
        };
        this.cameraController = new CameraController(options, true);
        this.accessory.configureController(this.cameraController);

        // Set the doorbell as the primary service
        if (this.accessory._associatedHAPAccessory) {
            this.accessory._associatedHAPAccessory
                .setPrimaryService(this.doorbellService);
        }

        // Link services
        this.speakerService.addLinkedService(this.controlService);
        this.microphoneService.addLinkedService(this.controlService);

        // No recent event triggers
        this.recentTriggers = {};

        // Register for status updates from the SkyBell device
        skybellDevice.setOptions({
            callbackInfo:     this.updateInfo.bind(this),
            callbackSettings: this.updateSettings.bind(this),
            callbackButton:
                activity => { this.trigger('cloud', 'button', activity) },
            callbackMotion:
                activity => { this.trigger('cloud', 'motion', activity) }
        });

        // Register for webhooks if configured
        if (webhooks) {
            webhooks.addHook('trigger/button', { name: skybellDevice.name },
                             () => { this.trigger('webhook', 'button') });
            webhooks.addHook('trigger/motion', { name: skybellDevice.name },
                             () => { this.trigger('webhook', 'motion') });
        }
    }

    // The device's information has been updated
    updateInfo(info) {
        this.log("updateInfo '" + this.name + "': serial=" + info.serialNo
                 + ', firmware=' + info.firmwareVersion
                 + ', hardware=' + info.hardwareRevision);
        
        // Set the values of dynamic characteristics
        this.informationService
            .updateCharacteristic(Characteristic.SerialNumber,
                                  info.serialNo)
            .updateCharacteristic(Characteristic.FirmwareRevision,
                                  info.firmwareVersion)
            .updateCharacteristic(Characteristic.Version,
                                  info.hardwareRevision);
    }

    // The device's settings have been updated
    updateSettings(settings) {
        this.log("updateSettings '" + this.name + "':"
                 + ' do_not_disturb=' + settings.do_not_disturb
                 + ', motion_policy=' + settings.motion_policy);

        // Map indoor chime and motion sensor enabled to HomeKit characteristics
        let indoorChimeActive = settings.do_not_disturb == 'true' ? 0 : 1;
        let motionSensorActive = settings.motion_policy == 'call' ? 1 : 0;

        // Map outdoor chime level to a HomeKit characteristic value
        let chimeVolume = settings.chime_level * 33;
        
        // Map LED brightness to HomeKit characteristic values
        let ledBrightness = Math.round(settings.led_intensity * 0.03) * 33;

        // Map LED colour to hue and saturation for HomeKit characteristics
        let rgb = [settings.green_r, settings.green_g, settings.green_b];
        let minRgb = Math.min(...rgb);
        let maxRgb = Math.max(...rgb);
        let chroma = maxRgb - minRgb;
        let hue;
        if (chroma == 0) {
            hue = 0; // (dummy value for white, i.e. R=G=B=255)
        } else if (maxRgb == rgb[0]) { // 0-60° or 300-360°
            hue = (rgb[1] - rgb[2]) / chroma;
            if (hue < 0) hue += 6;
        } else if (maxRgb == rgb[1]) { // 60-180°
            hue = (rgb[2] - rgb[0]) / chroma + 2;
        } else { // (maxRgb == rgb[2])    180-300°
            hue = (rgb[0] - rgb[1]) / chroma + 4;
        }
        let ledHue = Math.round(hue * 60);
        let ledSaturation = Math.round((chroma / maxRgb) * 100);
        
        // Set the values of the chime and LED characteristics
        this.doorbellService
            .updateCharacteristic(Characteristic.Active,     indoorChimeActive)
            .updateCharacteristic(Characteristic.Volume,     chimeVolume)
            .updateCharacteristic(Characteristic.Brightness, ledBrightness)
            .updateCharacteristic(Characteristic.Hue,        ledHue)
            .updateCharacteristic(Characteristic.Saturation, ledSaturation);
        
        // Set the value of the motion sensor enabled characteristic
        this.motionSensorService
            .updateCharacteristic(Characteristic.Active, motionSensorActive);

        // Set the maximum camera resolution
        let height = [1080, 720, 720, 480][settings.video_profile];
        this.streamingDelegate.setResolution(height);

        // Publish a new list of supported resolutions (this is a hack!)
        let options = this.streamingDelegate.getCodecParameters();
        this.cameraController.streamManagements.forEach(management => {
            let makeVideoStreamConfiguration =
                management._supportedVideoStreamConfiguration
                || management.constructor._supportedVideoStreamConfiguration;
            management.supportedVideoStreamConfiguration =
               makeVideoStreamConfiguration(options.video);
            management.service.setCharacteristic(
                Characteristic.SupportedVideoStreamConfiguration,
                management.supportedVideoStreamConfiguration);
        });
    }

    // Identify the SkyBell by flashing its LED (magenta and yellow)
    identify(paired, callback) {
        // Ignore if identify already running
        if (this.hasOwnProperty('identify')) {
            this.log("identify '" + this.name
                     + "' already in progress (ignored)");
            return callback();
        }

        // Start the animation
        this.log("identify '" + this.name + "'");
        this.identify = {
            callbacks: [callback],
            settings: [{
                led_intensity: this.skybellDevice.cache.settings.led_intensity,
                green_r:       this.skybellDevice.cache.settings.green_r,
                green_g:       this.skybellDevice.cache.settings.green_g,
                green_b:       this.skybellDevice.cache.settings.green_b
            }],
            interval: 1000
        };
        for (let i = 0; i < 2; ++i) {
            this.identify.settings.push({
                led_intensity: 100,
                green_r:       100,
                green_g:       0,
                green_b:       100
            });
            this.identify.settings.push({
                led_intensity: 100,
                green_r:       100,
                green_g:       100,
                green_b:       0
            });
        }
        this.identifyAnimate();
    }

    // Next step in identifying the SkyBell
    identifyAnimate() {
        // Set the next LED colour
        this.log("identifyAnimate '" + this.name + "': steps="
                 + this.identify.settings.length);
        this.identify.timer = null;
        let startTime = Date.now();
        this.skybellDevice.updateSettings(this.identify.settings.pop(),
                                          err => {
            if (!err && this.identify.settings.length) {
                // Schedule the next animation step
                let elapsed = Date.now() - startTime;
                let interval = Math.max(0, this.identify.interval - elapsed);
                this.log("identifyAnimate '" + this.name
                         + "' took=" + elapsed + 'ms'
                         + ', rescheduling in ' + interval + 'ms');
                this.identify.timer =
                    setTimeout(this.identifyAnimate.bind(this), interval);
            } else {
                // All done, so call the callback(s)
                this.log("identifyAnimate '" + this.name + "' finished");
                this.identify.callbacks.forEach(callback => callback(err));
                delete this.identify;
            }
        });
    }

    // Stop identifying the SkyBell
    identifyStop(callback) {
        if (this.hasOwnProperty('identify')) {
            // Terminate the identify animation as soon as possible
            if (this.identify.settings.length) {
                this.identify.settings.length = 1;
            }
            this.identify.interval = 0;
            this.identify.callbacks.push(callback);
            if (this.identify.timer) {
                clearTimeout(this.identify.timer);
                this.identify.timer =
                    setTimeout(this.identifyAnimate.bind(this));
            }
        } else {
            callback();
        }
    }

    // Set the state of the indoor chime (do not disturb)
    setIndoorChime(value, callback) {
        let settings = {
            do_not_disturb: value ? 'false' : 'true'
        };

        this.log("setIndoorChime '" + this.name + "': active=" + value
                 + ' as do_not_disturb=' + settings.do_not_disturb);
        this.skybellDevice.updateSettings(settings, callback);
    }

    // Set the volume of the outdoor chime
    setOutdoorChime(value, callback) {
        let settings = {
            chime_level: Math.round(value / 33)
        };

        this.log("setOutdoorChime '" + this.name + "': volume=" + value
                 + ' as level=' + settings.chime_level);
        this.skybellDevice.updateSettings(settings, callback);
    }

    // Set the brightness of the LED
    setLedIntensity(value, callback) {
        let settings = {
            led_intensity: [0, 25, 62, 100][Math.round(value / 33)]
        };
        
        this.log("setLedIntensity '" + this.name + "': brightness=" + value
                 + ' as intensity=' + settings.led_intensity);
        this.identifyStop(() => {
            this.skybellDevice.updateSettings(settings, callback);
        });
    }

    // Set the hue of the LED
    setLedHue(value, callback) {
        let saturation =
            this.doorbellService
                .getCharacteristic(Characteristic.Saturation).value;
        this.setLedColour(value, saturation, callback);
    }

    // Set the saturation of the LED
    setLedSaturation(value, callback) {
        let hue =
            this.doorbellService
                .getCharacteristic(Characteristic.Hue).value;
        this.setLedColour(hue, value, callback);
    }

    // Set the colour of the LED
    setLedColour(hue, saturation, callback) {
        // Convert the colour to RGB for the SkyBell settings
        let maxRgb = 255;
        let chroma = maxRgb * saturation / 100;
        let minRgb = maxRgb - chroma;
        let deltaRgb = chroma * ((hue / 60) % 1);
        let rgb;
        if (hue < 60) {
            rgb = [maxRgb, minRgb + deltaRgb, minRgb];
        } else if (hue < 120) {
            rgb = [maxRgb - deltaRgb, maxRgb, minRgb];
        } else if (hue < 180) {
            rgb = [minRgb, maxRgb, minRgb + deltaRgb];
        } else if (hue < 240) {
            rgb = [minRgb, maxRgb - deltaRgb, maxRgb];
        } else if (hue < 300) {
            rgb = [minRgb + deltaRgb, minRgb, maxRgb];
        } else { // (h < 360)
            rgb = [maxRgb, minRgb, maxRgb - deltaRgb];
        }
        let settings = {
            green_r: Math.round(rgb[0]),
            green_g: Math.round(rgb[1]),
            green_b: Math.round(rgb[2]),
        };

        this.log("setLedColor '" + this.name + "': hue=" + hue
                 + ', saturation=' + saturation
                 + ' as RGB=(' + settings.green_r + ', ' + settings.green_g
                 + ', ' + settings.green_b + ')');
        this.identifyStop(() => {
            this.skybellDevice.updateSettings(settings, callback);
        });
    }

    // Set the motion sensor policy
    setMotionPolicy(value, callback) {
        // Map the motion policy for the SkyBell settings
        let settings = {
            motion_policy: value ? 'call' : 'disabled'
        };
        
        this.log("setMotionPolicy '" + this.name + "': active=" + value
                 + ' as motion_policy=' + settings.motion_policy);
        this.skybellDevice.updateSettings(settings, callback);
    }
    
    // The button has been pressed
    buttonPressed() {
        this.log("buttonPressed '" + this.name + "'");
        this.doorbellService
            .updateCharacteristic(Characteristic.ProgrammableSwitchEvent,
                                  SINGLE_PRESS);
    }

    // Motion has been detected
    motionDetected() {
        this.log("motionDetected '" + this.name + "'");
        this.motionSensorService
            .updateCharacteristic(Characteristic.MotionDetected, true);

        // End the motion detection after a fixed delay
        clearTimeout(this.timerMotionEnd);
        this.timerMotionEnd = setTimeout(() => {
            this.log("motionDetected '" + this.name + "' finished");
            this.motionSensorService
                .updateCharacteristic(Characteristic.MotionDetected, false);
        }, DURATION_MOTION);
    }

    // Set the last activity
    setLastActivity(activity) {
        this.lastActivity = activity;

        // Set the activity to replay instead of streaming live video
        this.streamingDelegate.setActivity(activity);
    }

    // A button press or motion event trigger has been received
    trigger(source, type, activity) {
        // Remeber the most recent activity (of any type) for a short period
        if (activity) {
            this.log("trigger '" + this.name + "': Remembering activity");
            this.setLastActivity(activity);
            clearTimeout(this.timerLastActivity);
            this.timerLastActivity = setTimeout(() => {
                this.log("trigger '" + this.name + "': Forgetting activity");
                this.setLastActivity(null);
            }, DURATION_RECORDED);
        }

        // Check whether the event has come from another source recently
        let recent = this.recentTriggers[type];
        if (recent && (source != recent.source)) {
            return this.log("trigger '" + this.name + "': Suppressing "
                            + type + ' event from ' + source
                            + ' due to recent trigger from ' + recent.source);
        }

        // Suppress other sources of this event for a short period
        if (recent) clearTimeout(recent.timer);
        this.recentTriggers[type] = {
            source: source,
            timer:  setTimeout(() => {
                        this.log("trigger '" + this.name + "': Re-enabling "
                                 + type + ' triggers from all sources');
                        this.recentTriggers[type] = null;
                    }, DURATION_SUPPRESS)
        };

        // Process this event trigger
        if (type == 'button') {
            this.buttonPressed();
        } else if (type == 'motion') {
            this.motionDetected();
        }
    }
}
